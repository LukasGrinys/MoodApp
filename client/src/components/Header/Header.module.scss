@import '../../styles/config/variables';

.header {
    color: $color-night-white;
    width: 100%;
    position: sticky;
    top: 0;
    left: 0;
    display: flex;
    flex-direction: column;
    z-index: 999;

    .top {
        position: relative;
        background: $color-primary;
        box-shadow: $shadow-header;
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        align-items: center;
        padding: 8px;
        z-index: 99;

        &.nightMode {
            background: $color-night-darkgrey;
        }

        .logo {
            position: absolute;
            top: 4px;
            font-family: "Kaushan Script";
            font-size: 1.4em;
            cursor: default;
        }
        
        .bars {
            margin: 0 auto;
            cursor: pointer;
            animation: rotateButton 2s;
            transition: 1s;
        }
    }
}

.nightmode_button {
    width: 30px;
    opacity: 0.7;
    cursor: pointer;
    font-size: 25px;
    position: absolute;
    top: 5px;
    right: 0;
    transition-duration: 0.3s;

    &:hover {
        opacity: 1;

        & > .nightmode_button_caption {
            opacity: 1;
        }
    }

    .nightmode_button_caption {
        opacity: 0;
        position: absolute;
        top: 10px;
        right: 35px;
        width: 70px;
        font-size: 11px;
        transition-duration: 0.3s;
        text-align: right;
    }
}
