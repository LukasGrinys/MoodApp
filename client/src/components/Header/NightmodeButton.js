import React from 'react';
import styles from './Header.module.scss';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSun, faMoon } from '@fortawesome/free-solid-svg-icons';
import { useThemeUpdate, useTheme } from '../../contexts/ThemeContext';

const NightmodeButton = () => {
    const darkTheme = useTheme();
    const updateTheme = useThemeUpdate();

    return (
        <button 
            className={styles.nightmode_button} 
            onClick={updateTheme}
        >
            <FontAwesomeIcon 
                icon={darkTheme ? faSun : faMoon} 
                style={{fontSize:"1.5rem"}}
            />
            <div className={styles.nightmode_button_caption}> 
                {darkTheme ? `Day mode` : `Night mode`}
            </div>
        </button>
    );
};

export default NightmodeButton;