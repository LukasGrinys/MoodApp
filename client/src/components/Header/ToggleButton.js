import React from 'react';
import FontAwesome from 'react-fontawesome';
import styles from './Header.module.scss';

const ToggleButton = ({showNav, toggleNav}) => {
    return (
        <div className={styles.bars}>
            {
                 showNav ?
                 <FontAwesome
                     name="close"
                     onClick={toggleNav}
                     style={{
                         fontSize: '1.5em',
                         padding: "2px",
                         color: "#f2f2f2",
                         transform: "rotate(90deg)",
                         transition: "0.5s"
                     }}
                 />
             :
                 <FontAwesome
                     name="bars"
                     onClick={toggleNav}
                     style={{
                         fontSize: '1.5em',
                         padding: "2px",
                         color: "#f2f2f2",
                         transform: "rotate(180deg)",
                         transition: "0.5s"
                     }}
                 />
            }
        </div>
       
    );
};

export default ToggleButton;