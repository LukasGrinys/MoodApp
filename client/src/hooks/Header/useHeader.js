import { useState, useCallback } from 'react';

export const useHeader = () => {
    const [ showNav, setShowNav ] = useState(false);

    const toggleNav = useCallback(() => {
        setShowNav(!showNav)
    }, [showNav]);

    return {
        showNav,
        toggleNav
    }
}