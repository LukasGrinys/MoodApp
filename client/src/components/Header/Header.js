import React from 'react';
import styles from './Header.module.scss';
import { connect } from 'react-redux';
import NavBar from '../Navbar/NavBar';
import ToggleButton from './ToggleButton';
import NightmodeButton from './NightmodeButton';
import classNames from 'classnames';
import { useHeader } from '../../hooks/Header/useHeader';
import { useThemeUpdate, useTheme } from '../../contexts/ThemeContext';

const Header = ( {isAuth} ) => {
    const darkTheme = useTheme();
    const updateTheme = useThemeUpdate();
    const {
        showNav,
        toggleNav
    } = useHeader();

    return (
        <div className={styles.header}>
            <div 
                className={classNames(styles.top, darkTheme && styles.nightMode)} 
                style={darkTheme ? {backgroundColor: "#2F2F2F"} : {}}
            >
                <div className={styles.logo}>
                    Moodapp
                </div>
                <ToggleButton
                    showNav={showNav}
                    toggleNav={toggleNav}
                />
                <NightmodeButton
                    updateTheme={updateTheme}
                    darkTheme={darkTheme}
                />
            </div>
            {showNav && <NavBar toggleNav={toggleNav} isAuth={isAuth}/>}
        </div>
    );
}

function mapStateToProps({user}) {
    const { isAuth } = user;

    return { isAuth }
}


export default connect(mapStateToProps)(Header);