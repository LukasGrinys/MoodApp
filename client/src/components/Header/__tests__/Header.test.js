import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import Header from '../Header';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';

const mockStore = configureStore();
const store = mockStore({
    user: {
        isAuth: false
    }
});

const wrapper = mount(
    <Router>
        <Provider store={store}>
            <Header/>
        </Provider>
    </Router>
)

describe('Initial header UI', () => {
    test('Renders properly', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    test('Has logo', () => {
        expect(wrapper.find('.logo').length).toBe(1);
    });
    
    test('Has toggle navigation button', () => {
        expect(wrapper.find('ToggleButton').length).toBe(1);
    });
    
    test('Has nighmode button', () => {
        expect(wrapper.find('NightmodeButton').length).toBe(1);
    })
});

describe('Opening and closing navigation', () => {
    test('Opens nav when clicked on Bar icon', () => {
        expect(wrapper.find('NavBar').length).toBe(0);

        wrapper.find('FontAwesome').at(0).simulate('click');
        wrapper.update();

        expect(wrapper.find('NavBar').length).toBe(1);
    });

    test('Navigation button icon changes to close', () => {
        expect(wrapper.find('FontAwesome').at(0).prop('name')).toBe('close');
    });

    test('Closes nav when clicked on Close icon', () => {
        expect(wrapper.find('NavBar').length).toBe(1);

        wrapper.find('FontAwesome').at(0).simulate('click');
        wrapper.update();

        expect(wrapper.find('NavBar').length).toBe(0);
    });

    test('Navigation button icon changes to bars', () => {
        expect(wrapper.find('FontAwesome').at(0).prop('name')).toBe('bars');
    });
});

describe('Toggling nightmode', () => {
    test('Toggles nightmode when clicked', () => {
        expect(wrapper.find('.top').prop('className')).toBe('top');

        wrapper.find('NightmodeButton').simulate('click');
        wrapper.update();

        setTimeout( () => {
            expect(wrapper.find('.top').prop('className')).toBe('top nightMode');
            expect(wrapper.find('svg').prop('data-icon')).toBe('sun');
            done();
        })
    });

    test('Toggles nightmode off when clicked', () => {
        // Toggle it on first
        wrapper.find('NightmodeButton').simulate('click');
        wrapper.update();

        setTimeout( () => {
            expect(wrapper.find('.top').prop('className')).toBe('top nightMode');

            // Toggle it off then
            wrapper.find('NightmodeButton').simulate('click');
            wrapper.update();

            setTimeout( () => {
                expect(wrapper.find('.top').prop('className')).toBe('top');
                expect(wrapper.find('svg').prop('data-icon')).toBe('moon');
                done();
            })
        }, 0)
    });
})
