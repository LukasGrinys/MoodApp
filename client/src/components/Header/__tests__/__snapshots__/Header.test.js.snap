// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initial header UI Renders properly 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Connect(Header)>
        <Header
          dispatch={[Function]}
          isAuth={false}
        >
          <div
            className="header"
          >
            <div
              className="top"
              style={Object {}}
            >
              <div
                className="logo"
              >
                Moodapp
              </div>
              <ToggleButton
                showNav={false}
                toggleNav={[Function]}
              >
                <div
                  className="bars"
                >
                  <FontAwesome
                    name="bars"
                    onClick={[Function]}
                    style={
                      Object {
                        "color": "#f2f2f2",
                        "fontSize": "1.5em",
                        "padding": "2px",
                        "transform": "rotate(180deg)",
                        "transition": "0.5s",
                      }
                    }
                  >
                    <span
                      aria-hidden={true}
                      className="fa fa-bars"
                      onClick={[Function]}
                      style={
                        Object {
                          "color": "#f2f2f2",
                          "fontSize": "1.5em",
                          "padding": "2px",
                          "transform": "rotate(180deg)",
                          "transition": "0.5s",
                        }
                      }
                    />
                  </FontAwesome>
                </div>
              </ToggleButton>
              <NightmodeButton>
                <button
                  className="nightmode_button"
                >
                  <FontAwesomeIcon
                    border={false}
                    className=""
                    fixedWidth={false}
                    flip={null}
                    icon={
                      Object {
                        "icon": Array [
                          512,
                          512,
                          Array [],
                          "f186",
                          "M283.211 512c78.962 0 151.079-35.925 198.857-94.792 7.068-8.708-.639-21.43-11.562-19.35-124.203 23.654-238.262-71.576-238.262-196.954 0-72.222 38.662-138.635 101.498-174.394 9.686-5.512 7.25-20.197-3.756-22.23A258.156 258.156 0 0 0 283.211 0c-141.309 0-256 114.511-256 256 0 141.309 114.511 256 256 256z",
                        ],
                        "iconName": "moon",
                        "prefix": "fas",
                      }
                    }
                    inverse={false}
                    listItem={false}
                    mask={null}
                    pull={null}
                    pulse={false}
                    rotation={null}
                    size={null}
                    spin={false}
                    style={
                      Object {
                        "fontSize": "1.5rem",
                      }
                    }
                    swapOpacity={false}
                    symbol={false}
                    title=""
                    transform={null}
                  >
                    <svg
                      aria-hidden="true"
                      className="svg-inline--fa fa-moon fa-w-16 "
                      data-icon="moon"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      style={
                        Object {
                          "fontSize": "1.5rem",
                        }
                      }
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M283.211 512c78.962 0 151.079-35.925 198.857-94.792 7.068-8.708-.639-21.43-11.562-19.35-124.203 23.654-238.262-71.576-238.262-196.954 0-72.222 38.662-138.635 101.498-174.394 9.686-5.512 7.25-20.197-3.756-22.23A258.156 258.156 0 0 0 283.211 0c-141.309 0-256 114.511-256 256 0 141.309 114.511 256 256 256z"
                        fill="currentColor"
                        style={Object {}}
                      />
                    </svg>
                  </FontAwesomeIcon>
                  <div
                    className="nightmode_button_caption"
                  >
                    Night mode
                  </div>
                </button>
              </NightmodeButton>
            </div>
          </div>
        </Header>
      </Connect(Header)>
    </Provider>
  </Router>
</BrowserRouter>
`;
